### ISSUE TITLE
Make bitcoin-hashes optional, use traits/generics for hashes, so a library user can insert another implementation

### ISSUE THREAD

**Issue by benma** [rust-bitcoin/rust-bitcoin#647]
When using this library in an embedded environment such as the BitBox02, binary size matters. In the BitBox02 firmware, there are already multiple implementations of various hashing functions. It would be great to use traits and generics where hashing is needed so we can reuse the existing implementations instead of pulling in another one, needlessly increasing the binary size.

**Comment by Kixunil** [comment]
I don't think you can make `bitcoin_hashes` optional since it defines newtypes, not just implements hashing functions. So if you want trait-based approach it needs to be in the `bitcoin_hashes` crate and should actually be discussed over there.

**Comment by apoelstra** [comment]
I definitely like this in principle but it does sound like a hard cross-crate coordination problem.

Even within this library we have multiple sha2 implementations since there is one hidden in side libsecp256k1.

**Comment by thomaseizinger** [comment]
> So if you want trait-based approach it needs to be in the `bitcoin_hashes` crate and should actually be discussed over there.

Not necessarily. `rust-bitcoin` could define the traits it needs to provide further functionality and implement them upon certain dependency-features (like `bitcoin_hashes`) being activated.

```rust
pub trait HashToMessage {
	fn hash_sha256(data: &[u8]) -> Message;
}

#[cfg(feature = "bitcoin_hashes")]
pub struct BitcoinHashes;

#[cfg(feature = "bitcoin_hashes")]
impl HashToMessage for BitcoinHashes {
	// use bitcoin_hashes functions within trait implementations
}
```

Any function that needs hashing can the be parameterized over `T: HashToMessage`. With the traits being defined within `rust-bitcoin`, users could supply their own implementations if they wanted to.

This would essentially be a ports-and-adapter / onion / hexagonal architecture.

**Comment by Kixunil** [comment]
@thomaseizinger how would you combine this with newtypes found in `bitcoin_hashes` without changing them?

**Comment by thomaseizinger** [comment]
Probably best to strip the newtypes from `bitcoin_hashes` and have it concerned with the algo implementation only. In other words, expose function signatures that go from bytes to bytes.

If we want newtypes (like `PubkeyHash`), then those are probably more suited to live in rust-bitcoin which knows about the domain of public keys.

Then the `pubkey_hash` function of `PublicKey` can return a newtype and still be generic over the hash implementation.

**Comment by apoelstra** [comment]
@thomaseizinger the hashes do live in rust-bitcoin (they are created with macros from `bitcoin_hashes` but these are easy to move to rust-bitcoin if need be). The issue, as I understand it, is that they are newtypes around the concrete types from `bitcoin_hashes` and it is not obvious how we could genericize them based on alternate sha2 providers.

**Comment by thomaseizinger** [comment]
> It was my understanding that `bitcoin_hashes` is meant to separate primitives that are same across different applications. In that case removing newtypes would be a step back. Am I misunderstanding it?

I see them as implementations of hash functions used in Bitcoin ecosystem. The hash functions themselves are generic though. i.e. IMO nothing should stop someone to use the sha2 implementation if they need one.

Hence, I think a `&[u8] -> [u8; 32]` API or similar would probably be the most useful.

> @thomaseizinger the hashes do live in rust-bitcoin (they are created with macros from `bitcoin_hashes` but these are easy to move to rust-bitcoin if need be). The issue, as I understand it, is that they are newtypes around the concrete types from `bitcoin_hashes` and it is not obvious how we could genericize them based on alternate sha2 providers.

I think we would have to move away from macro-generated newtypes and instead do something like:

```rust
pub struct PubkeyHash(pub(crate) [u8; 20]);

pub trait Hash160 {
	fn hash(data: &[u8]) -> [u8; 20];
}

#[cfg(feature = "bitcoin_hashes")]
pub struct BitcoinHashes;

#[cfg(feature = "bitcoin_hashes")]
impl Hash160 for BitcoinHashes {
	// use bitcoin_hashes functions within trait implementations
}

impl PublicKey {
	pub fn pubkey_hash<H: Hash160>(&self) -> PubkeyHash {
		PubkeyHash(H::hash(...))	
	}
}
```

But that is not the most ergonomic ...

**Comment by dr-orlovsky** [comment]
MSRV increase to 1.36 & 2018 edition opens a way for doing normal derivation macros with syn (and helper crates like [`amplify_syn`](https://docs.rs/amplify_syn/1.1.6/amplify_syn/) can make this even more easier). This is a right way of getting rid of newtype macros. I have already quite a big experience in doing this derivation macros (for instance in [`amplify_derive`](https://crates.io/crates/amplify_derive) and [`lightning_encoding`](https://github.com/LNP-BP/lnp-core/blob/master/lightning_encoding/derive/tests/simple.rs) crates) and can do that part here.

**Comment by apoelstra** [comment]
I wouldn't mind a feature flag that would pass through to `sha2` if the user was willing to take the MSRV/stability hit.

Regarding generic traits, I wonder if we could just use constfns and associated constants to eliminate the code which is currently in a newtype macro.

**Comment by benma** [comment]
@apoelstra mentioned we could solve this with a `--cfg` flag and calling to extern "C" functions:

https://github.com/rust-bitcoin/rust-bitcoin/issues/2881#issuecomment-2226942945

This is probably the most viable solution, as it doesn't pollute the whole code with generics.

extern "Rust" might be even better as Rust users then don't have to deal with the restrictions of C functions, while users who need C functions can still call to C.

**Comment by Kixunil** [comment]
The issue with that is we still haven't figured out what to do about state. The only thing that sees doable to me is to allocate on heap.

**Comment by benma** [comment]
Yeah, for me that would be totally fine.
